File: ShadowNinjaGame.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.util.ArrayList;
import java.util.Random;

public class ShadowNinjaGame extends JFrame {
    private GamePanel gamePanel;

    public ShadowNinjaGame() {
        setTitle("Shadow Ninja: Trials of the Dojo");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setResizable(false);
        gamePanel = new GamePanel();
        add(gamePanel);
        pack();
        setLocationRelativeTo(null);
        setVisible(true);
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> new ShadowNinjaGame());
    }
}

class GamePanel extends JPanel implements KeyListener, Runnable {
    private static final int WIDTH = 800;
    private static final int HEIGHT = 600;
    private static final int FPS = 60;

    private Player player;
    private ArrayList<Enemy> enemies;
    private ArrayList<Platform> platforms;
    private ArrayList<Coin> coins;
    private ArrayList<Trap> traps;
    private boolean[] keys = new boolean[256];
    private Thread gameThread;

    public GamePanel() {
        setPreferredSize(new Dimension(WIDTH, HEIGHT));
        setBackground(Color.BLACK);
        setFocusable(true);
        addKeyListener(this);

        initializeGame();

        gameThread = new Thread(this);
        gameThread.start();
    }

    private void initializeGame() {
        player = new Player(100, 400);
        enemies = new ArrayList<>();
        enemies.add(new Enemy(300, 450));
        enemies.add(new Enemy(600, 250));

        platforms = new ArrayList<>();
        platforms.add(new Platform(0, 500, 800, 20));
        platforms.add(new Platform(200, 400, 200, 20));
        platforms.add(new Platform(500, 300, 200, 20));

        coins = new ArrayList<>();
        coins.add(new Coin(220, 380));
        coins.add(new Coin(520, 280));

        traps = new ArrayList<>();
        traps.add(new Trap(250, 480));
    }

    @Override
    public void run() {
        long lastTime = System.nanoTime();
        long frameTime = 1000000000 / FPS;

        while (true) {
            long currentTime = System.nanoTime();
            if (currentTime - lastTime >= frameTime) {
                update();
                repaint();
                lastTime = currentTime;
            }
        }
    }

    private void update() {
        player.update(keys, platforms, enemies);

        for (Enemy enemy : enemies) {
            enemy.update(player);
        }

        for (Trap trap : traps) {
            trap.checkCollision(player);
        }

        coins.removeIf(coin -> coin.checkCollision(player));
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        Graphics2D g2d = (Graphics2D) g;

        // Draw platforms
        g2d.setColor(Color.GRAY);
        for (Platform platform : platforms) {
            platform.draw(g2d);
        }

        // Draw coins
        g2d.setColor(Color.GREEN);
        for (Coin coin : coins) {
            coin.draw(g2d);
        }

        // Draw traps
        g2d.setColor(Color.RED);
        for (Trap trap : traps) {
            trap.draw(g2d);
        }

        // Draw player
        player.draw(g2d);

        // Draw enemies
        for (Enemy enemy : enemies) {
            enemy.draw(g2d);
        }

        // Draw UI
        g2d.setColor(Color.WHITE);
        g2d.setFont(new Font("Arial", Font.PLAIN, 18));
        g2d.drawString("Health: " + player.getHealth(), 10, 20);
        g2d.drawString("Coins: " + player.getCoins(), 10, 50);
    }

    @Override
    public void keyPressed(KeyEvent e) {
        if (e.getKeyCode() < 256) keys[e.getKeyCode()] = true;
    }

    @Override
    public void keyReleased(KeyEvent e) {
        if (e.getKeyCode() < 256) keys[e.getKeyCode()] = false;
    }

    @Override
    public void keyTyped(KeyEvent e) {}
}

class Player {
    private int x, y, width = 40, height = 60;
    private int velX = 0, velY = 0;
    private int health = 100, coins = 0;
    private boolean onGround = false, doubleJump = true;
    private boolean stealth = false;
    private int dashCooldown = 0;

    public Player(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public void update(boolean[] keys, ArrayList<Platform> platforms, ArrayList<Enemy> enemies) {
        velX = 0;
        if (keys[KeyEvent.VK_LEFT] || keys[KeyEvent.VK_A]) velX = -5;
        if (keys[KeyEvent.VK_RIGHT] || keys[KeyEvent.VK_D]) velX = 5;

        stealth = keys[KeyEvent.VK_S];

        if (keys[KeyEvent.VK_SPACE] && onGround) {
            velY = -15;
            onGround = false;
        } else if (keys[KeyEvent.VK_SPACE] && doubleJump && !onGround) {
            velY = -15;
            doubleJump = false;
        }

        velY += 1; // Gravity

        x += velX;
        y += velY;

        onGround = false;
        for (Platform platform : platforms) {
            if (x + width > platform.x && x < platform.x + platform.width &&
                y + height >= platform.y && y + height <= platform.y + 20) {
                y = platform.y - height;
                velY = 0;
                onGround = true;
                doubleJump = true;
            }
        }

        if (keys[KeyEvent.VK_F]) {
            for (Enemy enemy : enemies) {
                if (Math.abs(x - enemy.x) < 100) {
                    enemy.takeDamage(20);
                }
            }
        }

        if (dashCooldown > 0) dashCooldown--;
    }

    public void draw(Graphics2D g) {
        g.setColor(stealth ? new Color(0, 0, 255, 100) : Color.BLUE);
        g.fillRect(x, y, width, height);
    }

    public int getHealth() { return health; }
    public int getCoins() { return coins; }
    public void addCoin() { coins++; }
    public void takeDamage(int damage) { health -= damage; }
}

class Enemy {
    public int x, y, width = 40, height = 60;
    public int health = 50;
    private int visionCone = 100;
    private boolean alerted = false;

    public Enemy(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public void update(Player player) {
        if (Math.abs(x - player.x) < visionCone && !player.stealth) {
            alerted = true;
            if (x < player.x) x += 2;
            else x -= 2;
        } else {
            alerted = false;
            x += new Random().nextBoolean() ? 1 : -1;
        }
    }

    public void draw(Graphics2D g) {
        g.setColor(Color.RED);
        g.fillRect(x, y, width, height);
    }

    public void takeDamage(int damage) { health -= damage; }
}

class Platform {
    public int x, y, width, height;

    public Platform(int x, int y, int width, int height) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    public void draw(Graphics2D g) {
        g.fillRect(x, y, width, height);
    }
}

class Coin {
    public int x, y, size = 20;

    public Coin(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public boolean checkCollision(Player player) {
        if (player.x + 40 > x && player.x < x + size &&
            player.y + 60 > y && player.y < y + size) {
            player.addCoin();
            return true;
        }
        return false;
    }

    public void draw(Graphics2D g) {
        g.fillOval(x, y, size, size);
    }
}

class Trap {
    public int x, y, width = 40, height = 20;

    public Trap(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public void checkCollision(Player player) {
        if (player.x + 40 > x && player.x < x + width &&
            player.y + 60 > y && player.y < y + height) {
            player.takeDamage(10);
        }
    }

    public void draw(Graphics2D g) {
        g.fillRect(x, y, width, height);
    }
}

File: shadow_ninja_game.py
import pygame
import random
import math

# Initialize Pygame
pygame.init()

# Screen dimensions
SCREEN_WIDTH = 800
SCREEN_HEIGHT = 600
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
pygame.display.set_caption("Shadow Ninja: Trials of the Dojo")

# Colors
BLACK = (0, 0, 0)
WHITE = (255, 255, 255)
RED = (255, 0, 0)
GREEN = (0, 255, 0)
BLUE = (0, 0, 255)
GRAY = (128, 128, 128)

# Clock for FPS
clock = pygame.time.Clock()
FPS = 60

# Player class
class Player(pygame.sprite.Sprite):
    def __init__(self):
        super().__init__()
        self.image = pygame.Surface((40, 60))
        self.image.fill(BLUE)
        self.rect = self.image.get_rect()
        self.rect.x = 100
        self.rect.y = 400
        self.vel_x = 0
        self.vel_y = 0
        self.speed = 5
        self.jump_power = -15
        self.gravity = 0.8
        self.on_ground = False
        self.double_jump = True
        self.dash_cooldown = 0
        self.dash_speed = 15
        self.wall_jumping = False
        self.stealth = False
        self.health = 100
        self.coins = 0
        self.skills = {"double_jump": False, "dash": False}

    def update(self, platforms, enemies):
        keys = pygame.key.get_pressed()
        self.vel_x = 0
        if keys[pygame.K_LEFT] or keys[pygame.K_a]:
            self.vel_x = -self.speed
        if keys[pygame.K_RIGHT] or keys[pygame.K_d]:
            self.vel_x = self.speed
        if keys[pygame.K_SPACE] and self.on_ground:
            self.vel_y = self.jump_power
            self.on_ground = False
        elif keys[pygame.K_SPACE] and self.double_jump and not self.on_ground:
            self.vel_y = self.jump_power
            self.double_jump = False
        if keys[pygame.K_LSHIFT] and self.dash_cooldown == 0 and self.skills["dash"]:
            if keys[pygame.K_LEFT] or keys[pygame.K_a]:
                self.vel_x = -self.dash_speed
            elif keys[pygame.K_RIGHT] or keys[pygame.K_d]:
                self.vel_x = self.dash_speed
            self.dash_cooldown = 30
        if keys[pygame.K_s]:
            self.stealth = True
            self.image.set_alpha(100)
        else:
            self.stealth = False
            self.image.set_alpha(255)
        self.vel_y += self.gravity
        for platform in platforms:
            if self.rect.colliderect(platform.rect) and self.vel_x != 0:
                if self.vel_y > 0:
                    self.vel_y = 2
                    if keys[pygame.K_SPACE]:
                        self.vel_y = self.jump_power
                        self.vel_x = -self.vel_x
                        self.wall_jumping = True
        self.rect.x += self.vel_x
        self.rect.y += self.vel_y
        self.on_ground = False
        for platform in platforms:
            if self.rect.colliderect(platform.rect):
                if self.vel_y > 0:
                    self.rect.bottom = platform.rect.top
                    self.vel_y = 0
                    self.on_ground = True
                    self.double_jump = True
                    self.wall_jumping = False
                elif self.vel_y < 0:
                    self.rect.top = platform.rect.bottom
                    self.vel_y = 0
        if self.dash_cooldown > 0:
            self.dash_cooldown -= 1
        if keys[pygame.K_f]:
            for enemy in enemies:
                if self.rect.colliderect(enemy.rect):
                    enemy.health -= 20
        for coin in coins:
            if self.rect.colliderect(coin.rect):
                self.coins += 1
                coins.remove(coin)

class Enemy(pygame.sprite.Sprite):
    def __init__(self, x, y):
        super().__init__()
        self.image = pygame.Surface((40, 60))
        self.image.fill(RED)
        self.rect = self.image.get_rect()
        self.rect.x = x
        self.rect.y = y
        self.health = 50
        self.vision_cone = 100
        self.alerted = False

    def update(self, player):
        if abs(self.rect.x - player.rect.x) < self.vision_cone and not player.stealth:
            self.alerted = True
            if self.rect.x < player.rect.x:
                self.rect.x += 2
            else:
                self.rect.x -= 2
        else:
            self.alerted = False
            self.rect.x += random.choice([-1, 1])
        if self.health <= 0:
            self.kill()

class Platform(pygame.sprite.Sprite):
    def __init__(self, x, y, width, height):
        super().__init__()
        self.image = pygame.Surface((width, height))
        self.image.fill(GRAY)
        self.rect = self.image.get_rect()
        self.rect.x = x
        self.rect.y = y

class Trap(pygame.sprite.Sprite):
    def __init__(self, x, y):
        super().__init__()
        self.image = pygame.Surface((40, 20))
        self.image.fill(RED)
        self.rect = self.image.get_rect()
        self.rect.x = x
        self.rect.y = y

    def update(self, player):
        if self.rect.colliderect(player.rect):
            player.health -= 10

class Coin(pygame.sprite.Sprite):
    def __init__(self, x, y):
        super().__init__()
        self.image = pygame.Surface((20, 20))
        self.image.fill(GREEN)
        self.rect = self.image.get_rect()
        self.rect.x = x
        self.rect.y = y

all_sprites = pygame.sprite.Group()
platforms = pygame.sprite.Group()
enemies = pygame.sprite.Group()
traps = pygame.sprite.Group()
coins = pygame.sprite.Group()

player = Player()
all_sprites.add(player)

platform1 = Platform(0, 500, 800, 20)
platform2 = Platform(200, 400, 200, 20)
platform3 = Platform(500, 300, 200, 20)
platforms.add(platform1, platform2, platform3)
all_sprites.add(platform1, platform2, platform3)

enemy1 = Enemy(300, 450)
enemy2 = Enemy(600, 250)
enemies.add(enemy1, enemy2)
all_sprites.add(enemy1, enemy2)

trap1 = Trap(250, 480)
traps.add(trap1)
all_sprites.add(trap1)

coin1 = Coin(220, 380)
coin2 = Coin(520, 280)
coins.add(coin1, coin2)
all_sprites.add(coin1, coin2)

running = True
while running:
    clock.tick(FPS)
    screen.fill(BLACK)

    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False

    player.update(platforms, enemies)
    enemies.update(player)
    traps.update(player)

    all_sprites.draw(screen)

    font = pygame.font.SysFont(None, 24)
    health_text = font.render(f"Health: {player.health}", True, WHITE)
    coins_text = font.render(f"Coins: {player.coins}", True, WHITE)
    screen.blit(health_text, (10, 10))
    screen.blit(coins_text, (10, 40))

    if player.coins >= 2 and not player.skills["dash"]:
        player.skills["dash"] = True
        print("Dash unlocked!")

    pygame.display.flip()

pygame.quit()
